### ISSUE TITLE
Issues with long file names in FAT

### ISSUE THREAD

**Issue by RaulMerelli** [LTRData/DiscUtils#15]
First of all, thank you for maintaining the project.

I want to copy some files inside a FAT image, in read I see the long file names correctly.
When I go to write, however, it seems to fail to go beyond the 8.3 format, forcing the name to be uppercase (except the extension).

Below is an example of how I am operating, and how I then see the file by opening the image from 7zip:

```
string diskDump = "EFIESP.bin";
string sourcePath = @"DATA\developermenu.efi";
string destPath = @"Windows\System32\BOOT\developermenu.efi";

using (FileStream fs = File.Open(diskDump, FileMode.Open))
{
    if (!FatFileSystem.Detect(fs))
    {
        Console.WriteLine("Error: Dump is not formatted with FAT filesystem.");
        return;
    }
    using (FatFileSystem disk = new FatFileSystem(fs))
    {
        using (Stream s = disk.OpenFile(destPath, FileMode.Create))
        {
            using (var fileStream = File.Open(sourcePath, FileMode.Open))
            {
                fileStream.CopyTo(s);
            }
        }
    }
    Console.WriteLine("File copied.");
}
```
![image](https://github.com/user-attachments/assets/af992139-10ce-42b7-9763-b35920cfc90f)

I don't know if I did something wrong. By any chance are long names supported only in read and not in write? Thank you

**Comment by Olof-Lagerkvist** [comment]
That is correct, long names are not currently supported when creating new files. 

There is an old pull request in the old repository that solves long file name support in a better way to also support new files. I have plans to change to that solution at some point.

**Comment by xoofx** [comment]
Hey, I'm also looking for a good support for FAT partitions for similar use case (manipulate EFI system partitions)

> I have plans to change to that solution at some point.

What do you have in mind? I could give it a try to bring support for this long name issue (dunno how problematic currently things are) and make a PR if you are interested?

**Comment by Olof-Lagerkvist** [comment]
> Hey, I'm also looking for a good support for FAT partitions for similar use case (manipulate EFI system partitions)
> 
> > I have plans to change to that solution at some point.
> 
> What do you have in mind? I could give it a try to bring support for this long name issue (dunno how problematic currently things are) and make a PR if you are interested?

Sounds awesome!

Looking a bit deeper into it, it seems that https://github.com/DiscUtils/DiscUtils/pull/95 that I was referring to also only supported long file names for reading. It also seems to have some bugs that the current solution in this fork does not have. So, I would guess that the current solution over here is a better starting point after all.

What needs to be done is basically to create new LFN records in directories when new files are created with (or renamed to) a long file name. That is currently what is missing. Otherwise, LFNs are fully supported when listing directories and opening files using LFN.

**Comment by xoofx** [comment]
Great, I will start to have a look 🤞

**Comment by xoofx** [comment]
Ok, I had a first look (had to dig into whatever doc I could find about VFAT long names 😅), and it will take a bit of time to sort this out, but it seems doable. 

The current implementation for reading has actually subtle bugs (e.g it is not guaranteed that there is a `(char)0` null terminated character if the filename ends at exactly a multiple of 13, so it could potentially read garbage from the stackalloc), so it's good to take a 2nd look.

Hopefully, I should be able to introduce some optimizations as well (mostly reducing allocations). We will see. 🤞

**Comment by Olof-Lagerkvist** [comment]
> The current implementation for reading has actually subtle bugs (e.g it is not guaranteed that there is a `(char)0` null terminated character if the filename ends at exactly a multiple of 13, so it could potentially read garbage from the stackalloc), so it's good to take a 2nd look.

That should not be an issue as far as I can see. If it uses the entire record with 13 characters, no null character will be found (`IndexOf` returns -1) and the next line sets the recognized length to the entire length instead. It could never read any garbage characters this way.

**Comment by xoofx** [comment]
> That should not be an issue as far as I can see. If it uses the entire record with 13 characters, no null character will be found (`IndexOf` returns -1) and the next line sets the recognized length to the entire length instead. It could never read any garbage characters this way.

Oh, indeed, it was time to go to bed 😅

**Comment by gus33000** [comment]
Hi I would also be very interested in such additions, one of the projects I comaintained alongside the original author made use of a specific fork of DiscUtils (originally part of the repository itself) where we did modifications to support LFNs and fix a few bugs one of the PRs missed out, it is in no way perfect but I thought i would link to our changes. It would be amazing to see official support for it in this library fork.

Our forked version: https://github.com/gus33000/DiscUtils/commit/db6d16cdf3affb296d40a527035855627bc629ea

Commit db6d16cdf3affb296d40a527035855627bc629ea added the changes on top of the latest base off https://github.com/DiscUtils/DiscUtils

**Comment by xoofx** [comment]
Making some good progress, so sharing a quick update: ☺️ 

- I have implemented a new version of `FileName`, more lightweight (in terms of memory) and efficient, that also generates proper short name while handling collisions and tried to match the various behavior of a real FAT system. So far, it looks good (Not sure there are many implementations out there as accurate). The code is definitely more involving to handle all these cases, so I have added specific unit tests for testing this part without having to test the full FAT VFS, which should help to have more confidence.
- I should be able to proceed on plugging `FileName` ShortName/LongName into Directory resolution collisions. Might add some accelerating Dictionary to speed up the process (currently it is scanning all entries).
- A remaining challenge will be the proper handling of deleting LFN entries (that can span over multiple Directory entries) and reusing them. Currently the system is only able to handle 1 direntry per free list item, but I will need something that can handle multiple entries in order to avoid fragmentation.

I will probably need a couple of additional evenings to complete it. Learning some stuffs about FAT and LFN, so it's not that bad! 😅 

Stay tuned.
